### ISSUE TITLE
This code uses the empirical fisher matrix instead of actual fisher matrix.

### ISSUE THREAD

**Issue by siddharth9820** [gpauloski/kfac-pytorch#109]
### How did you install K-FAC and PyTorch?

```console
N/A
```

### What commit are you using?

N/A

### Describe the problem.

Hi,
First of all, great work! Thanks for putting together this useful code repository. 

As far as I understand, the actual fisher matrix is calculated from the gradients generated with **labels drawn from the network** and not the dataset. If one chooses to do the latter, it is called the empirical fisher matrix and there's a body of work which shows that using the empirical fisher matrix might not be a good idea for converfence. Even the original paper on K-FAC makes this distinction [1], (page 19, footnote). 

From what I can see the code uses the empirical fisher matrix instead of the actual one. Am I wrong in my understanding? If not, is there any reason for doing so?

The readme suggests that the code was built off of https://github.com/alecwangcq/KFAC-Pytorch, which uses labels generated from the network to calculate the fisher matrix (https://github.com/alecwangcq/KFAC-Pytorch/blob/master/main.py#L161-L167), thus being more faithful to the definition of Natural Gradients.

[1] Optimizing Neural Networks with Kronecker-factored Approximate Curvature, James Martens and Roger Grosse, URL - https://arxiv.org/pdf/1503.05671.pdf

**Comment by gpauloski** [comment]
Hi, @siddharth9820! Thanks for your interest in the project and for reaching out.

This is correct---we are using the empirical fisher matrix. While the K-FAC code was originally based on Chaoqi Wang's code, in our paper we evaluate against Kazuki Osawa's work (https://arxiv.org/pdf/1811.12019.pdf) which uses the empirical fisher (see the end of section 3.1).

Empirically, we did not find a measurable difference in convergence between the two methods, but that was quite a while ago and I think we only tested that with ResNet-32 and Cifar-10. That is to say, maybe the difference will matter more will larger models/different architectures.

If you add true fisher support and want to contribute it back, I'd be happy to work with you on a PR.
